webrtc.html

```diff
@@ -1620,8 +1623,80 @@ <h4>Set the RTCSessionDescription</h4>
                     lists.</p>
                   </li>
                   <li>
-                    <p>If <var>description</var> is set as a local description,
-                    then run the following steps:</p>
+                    <p>If <var>description</var> is of type <code>"rollback"</code>, then run
+                    the following steps:</p>
+                    <ol>
+                      <li data-tests="RTCPeerConnection-setDescription-transceiver.html">
+                        <p>If the <code><a data-link-for=
+                          "RTCRtpTransceiver">mid</a></code> value of an
+                        <code><a>RTCRtpTransceiver</a></code> was set to a
+                        non-null value by the
+                        <code><a>RTCSessionDescription</a></code> that is being
+                        rolled back, set the <code><a data-link-for=
+                          "RTCRtpTransceiver">mid</a></code> value of that
+                        transceiver to null, as described by <span data-jsep=
+                          "rollback">[[!JSEP]]</span>.</p>
+                      </li>
+                      <li data-tests="RTCPeerConnection-setDescription-transceiver.html">
+                        <p>If an <code><a>RTCRtpTransceiver</a></code> was
+                        created by applying the
+                        <code><a>RTCSessionDescription</a></code> that is
+                        being rolled back, and a track has not been attached to
+                        it via <code>addTrack</code>, remove that
+                        transceiver from <var>connection</var>'s <a
+                          href="#transceivers-set">set of transceivers</a>, as
+                        described by <span data-jsep=
+                          "rollback">[[!JSEP]]</span>.</p>
+                      </li>
+                      <li data-tests="RTCPeerConnection-setLocalDescription-rollback.html,RTCPeerConnection-setLocalDescription-answer.html,RTCPeerConnection-setRemoteDescription-rollback.html">
+                        <p>If <var>connection</var>.<a>[[\PendingLocalDescription]]</a>
+                        is not <code>null</code>, then set <var>description</var> to
+                        <var>connection</var>.<a>[[\CurrentLocalDescription]]</a>,
+                        and set <var>remote</var> to <code>false</code>. Otherwise,
+                        set <var>description</var> to
+                        <var>connection</var>.<a>[[\CurrentRemoteDescription]]</a>,
+                        and set <var>remote</var> to <code>true</code>.
```

Comment:
In other words "if we are rolling back a local description, set _description_ to the current local description, and if we're rolling back a remote description, set _description_ to the current remote description." And then we try to update to match _description_.

Just to make sure: Is this correct? Descriptions, both local and remote ones, can modify transceivers. Should we always roll back to a description based on the type of description being rolled back, or should we always roll back to the most up-to-date answer that got us to "stable"?

For example:
```
...
setLocalDescription(offer);
...
setRemoteDescription(answer);
setRemoteDescription(offer);
setRemoteDescription(rollback);
```

In this case we are rolling back a remote offer, and the current remote description would represent the answer SDP that got us to the previous "stable" state. That sounds good... I think?

```
...
setRemoteDescription(offer);  // 1)
...
setLocalDescription(answer);  // 2)
setRemoteDescription(offer);
setRemoteDescription(rollback);
```

But in this case, while we are also rolling back a remote offer, the current remote description would be the offer... however what was offered could have been rejected in the answer, and this was set as a _local_ description.

Assume offer 1) had a=sendrecv, but 2) made it a=recvonly because we didn't have anything to send. Wouldn't rolling back to the current remote description imply rolling back to 1), i.e. now we're firing "ontrack" again, even though currentDirection should become recvonly again (local answer), not sendrecv (remote offer).

@jan-ivar Could you shed some light on this? I'm not sure how trying to apply a particular current description is guaranteed to capture all the relevant states that need to be rolled back. Or what would happen if we modify transceivers while an offer is pending.

I might imagine that _description_ should be set to the "latest description applied that was an answer", but I'm not sure if that is correct either, e.g. does it always contain all the relevant m= sections or not?

Reply:
Ping @jan-ivar: Does the PR need updating or am I confused? (Or both :) )

Reply:
> Assume offer 1) had a=sendrecv, but 2) made it a=recvonly because we didn't have anything to send. Wouldn't rolling back to the current remote description imply rolling back to 1), i.e. now we're firing "ontrack" again, even though currentDirection should become recvonly again (local answer), not sendrecv (remote offer).

Thanks, yes you're right, I need to update the PR here. Our rejection API wort (expose in offer, revert in answer) is the gift that keeps on giving. 😕

We need to reconstruct not just from [[CurrentRemoteDescription]] in this case, but from [[CurrentDirection]] as well, to account for rejections.

We want to roll back correctly to recvonly in this case, and thankfully [[CurrentDirection]] has this value, thanks to existing prose in SLD(answer): *"Set transceiver.[[CurrentDirection]] and transceiver.[[FiredDirection]] to direction."*

As to your other questions, the scope of this PR is solely about firing the correct events, so I suggest we discuss those in the issue you opened.